src/rate.rs

```diff
@@ -0,0 +1,210 @@
+use std::convert::TryFrom;
+use std::str::FromStr;
+
+/// Represent a rate. Note this is designed to support Bitcoin/Dai buy and sell rates (Bitcoin being in the range of 10k-100kDai)
+// The rate is equal to integer * 10e-inv_dec_exp
+#[derive(Clone, Copy, Debug, PartialEq, Eq)]
+pub struct Rate {
+    integer: u32,
+    inv_dec_exp: usize,
+}
+
+impl Rate {
+    const MAX_PRECISION: usize = 9;
+
+    pub fn integer(&self) -> u32 {
+        self.integer
+    }
+
+    pub fn inverse_decimal_exponent(&self) -> usize {
+        self.inv_dec_exp
+    }
+
+    pub fn new(integer: u64, inverse_decimal_exponent: usize) -> anyhow::Result<Self> {
+        let (integer, inv_dec_exp) = Rate::reduce(integer, inverse_decimal_exponent);
+
+        let integer =
+            u32::try_from(integer).map_err(|_| anyhow::anyhow!("Value is unexpectedly large."))?;
+        Ok(Rate {
+            integer,
+            inv_dec_exp,
+        })
+    }
+
+    pub fn from_f64(rate: f64) -> anyhow::Result<Rate> {
+        if rate.is_sign_negative() {
+            anyhow::bail!("Spread must be positive");
+        }
+
+        if !rate.is_finite() {
+            anyhow::bail!("Spread must be finite")
+        }
+
+        let mut rate = rate.to_string();
+        let decimal_index = rate.find('.');
+        match decimal_index {
+            None => {
+                let rate = u64::from_str(&rate)
+                    .map_err(|_| anyhow::anyhow!("Rate is unexpectedly large."))?;
+                Rate::new(rate, 0)
+            }
+            Some(decimal_index) => {
+                let mantissa = rate.split_off(decimal_index + 1);
+                if mantissa.len() > Self::MAX_PRECISION {
+                    anyhow::bail!("Precision of the rate is too high.")
+                } else {
+                    // Removes the trailing decimal point
+                    rate.truncate(rate.len() - 1);
+                    let integer = rate;
+
+                    let inv_dec_exp = mantissa.len();
+
+                    let integer = u64::from_str(&format!("{}{}", integer, mantissa))
+                        .map_err(|_| anyhow::anyhow!("Rate is unexpectedly large"))?;
+                    Rate::new(integer, inv_dec_exp)
+                }
+            }
+        }
+    }
+
+    /// If the integer part ends with 0 and the inverse_decimal_exponent is not null
+    /// then we can reduce the representation by removing zero and decrementing the inverse
+    /// exponent. For example:
+    /// Rate { integer: 1000, inv_dec_exp: 1 } becomes
+    /// Rate { integer: 100, inv_dec_exp: 0 }.
+    fn reduce(integer: u64, inv_dec_exp: usize) -> (u64, usize) {
+        let mut integer_str = integer.to_string();
+        if integer_str.len() > 1 && integer_str.ends_with('0') && inv_dec_exp != 0 {
```

Comment:
```haskell
reduce :: String -> Integer -> (String, Integer)
reduce x 0 = (x, 0)
reduce x y 
  | (last x) == '0' = reduce (reverse (tail (reverse x))) (y - 1)
  | otherwise = (x, y)
```

Reply:
The recursion is already here but it could possibly be neater, I'll try.

Reply:
Removed :)